File: api/model/user/exam.go
package user

import (
	"time"

	"gorm.io/gorm"
)

// TestpaperRecord 考试记录
type TestpaperRecord struct {
	ID          uint           `gorm:"primaryKey" json:"id"`
	TestpaperID uint           `gorm:"not null;index" json:"testpaperId"`
	UserID      uint           `gorm:"not null;index" json:"userId"`
	CourseID    uint           `gorm:"not null;index" json:"courseId"`
	Score       int            `gorm:"default:0" json:"score"`  // 得分
	Passed      bool           `gorm:"default:false" json:"passed"` // 是否及格
	Answers     string         `gorm:"type:text" json:"answers"` // JSON: [{"questionId":1,"userAnswer":["A"],"correct":true,"score":10}]
	DraftAnswers string        `gorm:"type:text" json:"draftAnswers"` // 草稿答案 JSON: [{"questionId":1,"userAnswer":["A"]}]（交卷前自动保存，用于断点续考）
	StartedAt   time.Time      `json:"startedAt"`
	SubmittedAt *time.Time     `json:"submittedAt"`
	Duration    int            `gorm:"default:0" json:"duration"` // 实际用时（秒）
	CreatedAt   time.Time      `json:"createdAt"`
	UpdatedAt   time.Time      `json:"updatedAt"`
	DeletedAt   gorm.DeletedAt `gorm:"index" json:"-"`
}

// TableName 表名
func (TestpaperRecord) TableName() string { return "testpaper_records" }

File: front/src/api/exam.ts
import request from './request'
import type { ApiResult } from './request'

export interface ExamTestpaperItem {
  id: number
  name: string
  description: string
  type: number // 1随时考 2课程完成后考
  totalScore: number
  passScore: number
  duration: number // 分钟
  available: boolean
  hasFinished?: boolean // 是否有已完成的考试记录
  bestScore?: number // 最高分（hasFinished 为 true 时有效）
  bestPassed?: boolean // 最高分那次是否及格
  inProgress?: boolean // 是否有进行中的考试（未交卷）
}

export interface ExamQuestion {
  id: number
  type: number // 1单选 2多选 3判断
  title: string
  options: string // JSON
  score: number
  sort: number
}

export interface DraftAnswer {
  questionId: number
  userAnswer: string[]
}

export interface ExamDetail {
  id: number
  courseId: number
  name: string
  description: string
  type: number
  totalScore: number
  passScore: number
  duration: number
  remainSec: number // 剩余秒数
  questions: ExamQuestion[]
  draftAnswers: DraftAnswer[] // 草稿答案（断点续考用，新建时为空）
}

export interface AnswerDetail {
  questionId: number
  title: string
  type: number
  userAnswer: string[]
  correctAnswer: string[]
  correct: boolean
  score: number
  maxScore: number
}

export interface SubmitResult {
  score: number
  passed: boolean
  total: number
  passLine: number
  details: AnswerDetail[]
}

export interface ExamRecordItem {
  recordId: number
  testpaperId: number
  testpaperName: string
  score: number
  passed: boolean
  duration: number
  submittedAt: string
}

export const getCourseExam = (courseId: number) =>
  request.get<any, ApiResult<ExamTestpaperItem[]>>(`/user/course/${courseId}/exam`)

export const getExam = (testpaperId: number) =>
  request.get<any, ApiResult<ExamDetail>>(`/user/testpaper/${testpaperId}/exam`)

export const saveExamDraft = (testpaperId: number, answers: DraftAnswer[]) =>
  request.post<any, ApiResult<null>>(`/user/testpaper/${testpaperId}/draft`, { answers })

export const submitExam = (testpaperId: number, answers: { questionId: number; userAnswer: string[] }[]) =>
  request.post<any, ApiResult<SubmitResult>>(`/user/testpaper/${testpaperId}/submit`, { answers })

export const getExamRecords = (courseId: number) =>
  request.get<any, ApiResult<ExamRecordItem[]>>(`/user/course/${courseId}/exam/records`)
